src/lib/uri/uri.c

```diff
@@ -371,3 +372,92 @@ uri_set_create(struct uri_set *uri_set, const char *str)
 	uri_set->uri_count = 0;
 	return -1;
 }
+
+static unsigned char unreserved_symbols[256] = {0};
+
+/*
+ * Builds an array with unreserved characters.
+ */
+static void
+build_unreserved_symbols_table(const char *str)
+{
+	for (int i = 0; i < 256; i++)
+		unreserved_symbols[i] = 0;
+
+	for (; *str; str++) {
+		unsigned char ch = (unsigned char)*str;
+		unreserved_symbols[ch] = 1;
+	}
+}
+
+/*
+ * String percent-encoding.
+ */
+char *
+uri_escape(const char *buf, size_t len, const char *unreserved, bool encode_plus)
+{
+	build_unreserved_symbols_table(unreserved);
+	/* The worst case is when all characters are encoded. */
+	char *encoded = xcalloc(len * 3 + 1, sizeof(char));
+	int pos = 0;
+	const char *hex = "0123456789ABCDEF";
+	while (len--) {
+		unsigned char ch = (unsigned char)*buf;
+		if ((ch == ' ') && encode_plus) {
+			encoded[pos++] = '+';
+		} else if (!unreserved_symbols[ch]) {
+			encoded[pos++] = '%';
+			encoded[pos++] = hex[ch >> 4];
+			encoded[pos++] = hex[ch & 15];
+		} else {
+			encoded[pos++] = *buf;
+		}
+		buf++;
+	}
+	encoded[pos++] = '\0';
+	return encoded;
+};
+
+/*
+ * Converts a hex character to its integer value.
+ */
+static char
+hex2ch(char ch)
+{
+	return isdigit(ch) ? ch - '0' : tolower(ch) - 'a' + 10;
+}
+
+static int
+ishex(int x)
+{
+	return	(x >= '0' && x <= '9')	||
+		(x >= 'a' && x <= 'f')	||
+		(x >= 'A' && x <= 'F');
+}
+
+/*
+ * String percent-decoding.
+ */
+char *
+uri_unescape(const char *str, bool decode_plus)
+{
+	char *pstr = (char *)str;
+	char *buf = xmalloc(strlen(str) + 1);
+	char *pbuf = buf;
+	while (*pstr) {
+		if ((*pstr == '%') && (!ishex(pstr[1]))) {
+			*pbuf++ = '%';
+		} else if ((*pstr == '%') && (pstr[1]) && (pstr[2])) {
+			*pbuf++ = hex2ch(pstr[1]) << 4 | hex2ch(pstr[2]);
+			pstr += 2;
```

Comment:
The logic here attempts to decode `%<hex><hex>`, but leave `%<non-hex><non-hex>` unchanged. For example, `%@@` will remain `%@@`. It also handles `%<non-hex><hex>` correctly, but misses `%<hex><non-hex>`.

I think it is good to test:

1. `%<non-hex><non-hex>`
2. `%<hex><non-hex>`
3. `%<non-hex><hex>`
4. `%<EOS>` (`<EOS>` -- the end of a string).
5. `%<hex><EOS>`
6. `%<non-hex><EOS>`

Since we can't assume that the string ends with a zero byte (see other comments), we should explicitly check the position against the string size.

Some code to illustrate spoken points:

```c
char *
uri_unescape(const char *src, size_t src_size, bool plus, size_t *dst_size);
{
	char *dest = xmalloc(src_size * sizeof(char));
	char *dest_p = dest;
	const char *src_end = src + src_size;
	for (const char *p = src; p < src_end; ++p) {
		if (*p == '%') {
			bool is_hex_1 = p + 1 < src_end &&
				isxdigit((unsigned char)p[1]);
			bool is_hex_2 = p + 2 < src_end &&
				isxdigit((unsigned char)p[2]);
			if (is_hex_1 && is_hex_2) {
				*dest_p++ = hex2ch(p[1]) << 4 | hex2ch(p[2]);
				p += 2;
			} else {
				*dest_p++ = '%';
			}
		} else if (plus && *p == '+') {
			*dest_p++ = ' ';
		} else {
			*dest_p++ = *p;
		}
	}
	*dst_size = dest_p - dest;
	return dest;
}
```